0001-typedb/02-dev/07-response.md

```diff
@@ -0,0 +1,161 @@
+---
+
+# Response interpretation
+
+TypeDB can return a response in different formats, depending on a type of query. The response will be interpreted by a 
+TypeDB Client being used. TypeDB Client provides classes and methods for different response types respectively. Please 
+see the table below for mapping of response format to a type of query that has been sent.
+
+|  **#**  |             **Query type**              |          **Response format**           |
+|:-------:|:---------------------------------------:|:--------------------------------------:|
+|    1    |                Get query                |   Stream/Iterator of **ConceptMap**    |
+|    2    |       Get query with aggregation        |       **Future** of **Numeric**        |
+|    3    |         Get query with grouping         | Stream/Iterator of **ConceptMapGroup** |
+|    4    | Get query with grouping and aggregation |  Stream/Iterator of **NumericGroup**   |
+|    5    |              Insert query               |   Stream/Iterator of **ConceptMap**    |
+|    6    |              Delete query               |     **Future** (of empty response)     |
+|    7    |   Update (match-delete-insert) query    |   Stream/Iterator of **ConceptMap**    |
+|    8    |                 Define                  |     **Future** (of empty response)     |
+|    9    |                Undefine                 |     **Future** (of empty response)     |
+|   10    |                 Explain                 | Stream/Iterator of **Explanation**     |
+
+## Expected results
+
+See the table above for an overview of response formats. For more details see the [Response formats](#response-formats) 
+section below.
+
+### Get query
+
+The ordinary [get query](05-read.md#get-query) (without [aggregation](05-read.md#aggregation) or 
+[grouping](05-read.md#group)) returns a stream/iterator of a ConceptMap objects. 
+
+Every iteration returns result as a [ConceptMap](#conceptmap) object. 
+
+### Get query with aggregation
+
+The get query with **aggregation** can only return a singular value (aggregated result). Hence it returns a Future 
+object of Numeric.
+
+Use [Future](#future) object’s `get()` method to wait and retrieve the [Numeric](#numeric) object when it’s ready.
+
+### Get query with grouping
+
+The get query with **grouping** returns a stream/iterator of ConceptMapGroup objects.
+
+Every iteration should return result as a [ConceptMapGroup](#conceptmapgroup) object. 
+
+Use ConceptMapGroup object to retrieve [ConceptMap](#conceptmap).
+
+### Get query with grouping and aggregation
+
+The get query with **grouping** and **aggregation** returns a stream/iterator of a NumericGroup objects.
+
+Every iteration should return result as a [NumericGroup](#numericgroup) object.
+
+Use NumericGroup object to retrieve [Numeric](#numeric).
+
+### Insert query
+
+An insert query returns a stream/iterator of a ConceptMap objects. 
+
+Every iteration should return result as a [ConceptMap](#conceptmap) object. 
```

Comment:
I usually refer to them as `element in the iterator` but iteration may also be correct.

Reply:
Element is an object, but iteration is a process. Basically everything we do in context of a single element of iterator is iteration.
I do not see a problem here, but if you want I can delete the word `return`. So that process results in something. Seems a little bit more natural this way.

Reply:
nah, this is probably more correct.

Reply:
I guess I was influenced by places like this: https://wiki.python.org/moin/Iterator 
> An iterator object implements __next__, which is expected to return the next element of the iterable object that returned it, and to raise a StopIteration exception when no more elements are available.